### ISSUE TITLE
Cursor pagination doesn't work with search queries

### ISSUE THREAD

**Issue by benknight** [wp-graphql/wp-graphql#1818]
Cursor-based pagination combined with search queries generates SQL queries that do not correctly paginate search results. Investigating SQL Query Logs shows this is because cursor-based pagination translates to post date comparisons, which depends on queries being ordered by by post date. However, search queries are ordered by `wp_posts.post_title LIKE '%query%'` first, therefore causing the start/end cursors to reference posts that are not in chronological order.

During my investigation I found logic that was already put in place to prevent this, right here:
https://github.com/wp-graphql/wp-graphql/blob/develop/src/Data/Connection/PostObjectConnectionResolver.php#L262

```
/**
  * If the query contains search default the results to
  */
if ( isset( $query_args['search'] ) && ! empty( $query_args['search'] ) ) {
  /**
    * Don't order search results by title (causes funky issues with cursors)
    */

  $query_args['search_orderby_title'] = false;
  $query_args['orderby']              = 'date';
  $query_args['order']                = isset( $last ) ? 'ASC' : 'DESC';
}
```

However this block is never executing because it looks like the search query is at `$query_args['s']` not `$query_args['search']`.

**Example query:**

```graphql
query SearchPosts($after: String, $before: String, $query: String!) {
  posts(after: $after, before: $before, first: 10, last: null, where: {search: $query}) {
    pageInfo {
      startCursor
      endCursor
      hasNextPage
      hasPreviousPage
    }
    edges {
      cursor
    }
    nodes {
      databaseId
      title
      date
    }
  }
}
```

Will translate into the following SQL query:

`SELECT wp_posts.ID FROM wp_posts  WHERE 1=1  AND (((wp_posts.post_title LIKE '%example%') OR (wp_posts.post_excerpt LIKE '%example%') OR (wp_posts.post_content LIKE '%example%')))  AND wp_posts.post_type = 'post' AND ((wp_posts.post_status = 'publish')) AND  CAST( wp_posts.post_date as DATETIME ) <= CAST( '2014-06-03 12:19:21' as DATETIME ) AND ( CAST( wp_posts.post_date as DATETIME ) < CAST( '2014-06-03 12:19:21' as DATETIME ) OR (  wp_posts.ID < 10768 ) ) ORDER BY wp_posts.post_title LIKE '%example%' DESC, wp_posts.post_date DESC, wp_posts.ID DESC  LIMIT 0, 6
`

WPGraphQL version 1.3.3

**Comment by benknight** [comment]
Update - since updating the references, pagination indeed works now, however search results are awful now because it's no longer ordering results based on post title 'LIKE' query. It looks like a more robust solution is necessary to maintain both pagination and search result relevance.

**Comment by benknight** [comment]
Update - for now I plan to fall back to using the REST API to grab search results, then pass those IDs into a GraphQL query to get the post data I need.

https://developer.wordpress.org/rest-api/reference/search-results/

**Comment by jasonbahl** [comment]
@benknight Thanks for the details here. I'll look more into this. I think there's been some underlying fixes to pagination that might allow us to better support cursor pagination and different ordering. 

I'll look into this a bit more.

**Comment by jasonbahl** [comment]
@benknight ok, so I think the main culprit to the pagintation issues is this: https://github.com/WordPress/WordPress/blob/master/wp-includes/class-wp-query.php#L1496-L1550

It's always set to `DESC`, largely because WP doesn't have a concept of forward/backward pagination. 

So, I think for WPGraphQL, we'll have to have our own version of this `parse_search_order` function that we can apply with `ASC` and `DESC` order to properly support pagination _and_ relevant sorting/ordering. 

I think I see a path forward for this that hopefully won't take too long to implement 🤞

**Comment by markkelnar** [comment]
As I'm digging into this and trying to recreate with a sample set of posts, I'm seeing the same results in the same order between graphql, WP admin posts page and wp-json/wp/v2 rest api. I'm not able to recreate the problem or maybe I'm not looking in the right spot?  Can you compare what I have here with what you are doing or create a similar sample set of posts that I can use to recreate the problem?

Using graphql like this
`query SearchPosts($after: String, $before: String, $query: String!) {
  posts(after: $after, before: $before, first: 3, last: null, where: {search: $query}) {
    pageInfo {
      startCursor
      endCursor
      hasNextPage
      hasPreviousPage
    }
    edges {
      cursor
      node {
        databaseId
        title
        date
        content
      }
    }
  }
}
`

Querying per page at the api endpoint like this
`http://localhost:8091/wp-json/wp/v2/search?search=foo&per_page=3&page=1`

And in wp-admin 3 per page, all page results were the same.

![Screen Shot 2021-04-09 at 11 18 16 AM](https://user-images.githubusercontent.com/749603/114210977-b685e500-9925-11eb-9586-23b426baa594.png)

The sample posts exported from my WP.
[test.WordPress.2021-04-09.xml.zip](https://github.com/wp-graphql/wp-graphql/files/6287172/test.WordPress.2021-04-09.xml.zip)

**Comment by jasonbahl** [comment]
@benknight I paired with @markkelnar to try and work through this and we were not able to reproduce. If you're still experiencing this, could you maybe put together a screen recording showing how you're able to reproduce this?

**Comment by jasonbahl** [comment]
@benknight as we've tried to reproduce and haven't been able to and haven't seen an update from you, I'm going to close this. 

If you believe this is still an issue and can provide more information for us to reproduce, please open a new issue and reference this one for context. 

Thanks!

**Comment by benknight** [comment]
That's odd. I know I typed a very detailed response to this a week or two ago but now I don't see it here at all. Strange. 

Anyway, I believe this is still a bug and the analysis points to the exact lines of code and query that lead to unexpected results when paginating a posts query by search keywords.

If that's not enough already, let me know and I'll follow up with more details.

**Comment by benknight** [comment]
Here's a screen recording showing two result sets, where the second result set contains duplicated results from the first query:

https://user-images.githubusercontent.com/786691/125592374-f14f03b8-de47-4bac-a346-1461acdb6d8b.mov

**Comment by jasonbahl** [comment]
@benknight thanks for the follow-up. Bummer that the message didn't save to Github! 

I'll re-open and we can investigate again, but we might need to set a time to pair on this. @markkelnar and I weren't able to reproduce when we worked through this.

**Comment by tuanfront-end** [comment]
Cursor pagination will work not correct if in query have any order/orderBy filter. I have been checked this problem very carefully. Thanks

**Comment by benknight** [comment]
Just an update here, to be clear, the SQL query that is being generated is problematic. The crux of the issue is cursor pagination works by references post date ranges. However when adding `ORDER BY wp_posts.post_title LIKE '%example%' DESC` (see OP for details), this essentially breaks cursor pagination because the results are not ordered by date. This is a pretty clear cut reproducible issue and means paginated search results are completely broken and have been for months/years.

**Comment by benknight** [comment]
Sorry guys I just don't have any time to put into this. I thought I'd described why the SQL generated by is problematic but maybe I'm just crazy. It's odd that I'm the only person who's reported having this issue so many just close until someone else comes across this.

**Comment by jasonbahl** [comment]
Sounds good @benknight. I'm sure this is a bug, but I'm having a heck of a time reproducing it. 

I'll close this for now, and if similar issues are reported by other users we can open it up again. 

Thanks!

**Comment by tareq0065** [comment]
@jasonbahl any updates on this please? I am not getting any results for search either.

**Comment by justlevine** [comment]
@tareq0065 please open a new issue with an isolated test case

**Comment by devinatbryt** [comment]
Hrmmm, I was getting this exact same issue, unsure what the root of it was. That being said, my way around it was keeping track of all of the already queried paginated posts ids and along with a similar query above, included the `notIn` argument within where to filter out existing posts.

I hope this helps future devs who stumble into this very weird issue :smile: .

```
query MyQuery($search:String, $limit:Int, $after:String, $notIn:[ID]) {
  posts(first: $limit, after: $after, where: { search: $search, notIn: $notIn } ) {
    nodes {
      databaseId
      ...otherFields
    }
    pageInfo {
      endCursor
      startCursor
      hasNextPage
      hasPreviousPage
    }
  }
}
```

**Comment by justlevine** [comment]
@jasonbahl @markkelnar looking at this again, I think  #2550 is a duplicate. If so, I can help with the replication, but stabilizing the underlying sql/cursors might be a bit out of my range.

**Comment by jasonbahl** [comment]
@justlevine ya, they seem related. 

I wasn't able to reproduce when I tried. 

Glad you've figured out how to get it into a reproducible state! I'll check #2550. If I can reproduce it, I'm sure I can help fix it

**Comment by mboynes** [comment]
I encountered this issue and can provide an easy-to-replicate test case for you @jasonbahl. Recording it here instead of in #2550 as I think this issue more accurately describes what I'm observing. However, I suspect they are tightly coupled.

## Steps to reproduce

1. Spin up a site and load the [theme unit test data](https://github.com/WPTT/theme-test-data/blob/master/themeunittestdata.wordpress.xml) into it.
2. In the admin post list, search for "post format"to observe the expected results. You should see 19 results, the first 14 have "post format" in the title and the last 5 do not -- this is important. Note that the dates on the first 14 results are all in 2010, and the last 5 are 2012 and on -- also important.
3. In GraphiQL, execute the same search for "post format" and get the first 10 items. The first set of results aligns with our expectations. The second set (the IDs should have imported with the WXR, so this should be after `YXJyYXljb25uZWN0aW9uOjU1OQ==`) only contains the first 4 results of what we expect to see.

## Walkthrough

As @benknight noted, what's happening is because WPGraphQL inserts a date (or ID) clause into the sql query. Because of WordPress core's search results ordering, the 10th post's date is older than the last 5. The date clause in the SQL excludes these items (generated query below).

```sql
SELECT   wp_posts.ID
FROM wp_posts 
WHERE 1=1  AND (((wp_posts.post_title LIKE '%post%') OR (wp_posts.post_excerpt LIKE '%post%') OR (wp_posts.post_content LIKE '%post%')) AND ((wp_posts.post_title LIKE '%format%') OR (wp_posts.post_excerpt LIKE '%format%') OR (wp_posts.post_content LIKE '%format%')))  AND wp_posts.post_type = 'post' AND ((wp_posts.post_status = 'publish')) AND  CAST( wp_posts.post_date as DATETIME ) <= CAST( '2010-05-09 07:51:54' as DATETIME ) AND ( CAST( wp_posts.post_date as DATETIME ) < CAST( '2010-05-09 07:51:54' as DATETIME ) OR (  wp_posts.ID < 559 ) ) 
ORDER BY (CASE WHEN wp_posts.post_title LIKE '%post format%' THEN 1 WHEN wp_posts.post_title LIKE '%post%' AND wp_posts.post_title LIKE '%format%' THEN 2 WHEN wp_posts.post_title LIKE '%post%' OR wp_posts.post_title LIKE '%format%' THEN 3 WHEN wp_posts.post_excerpt LIKE '%post format%' THEN 4 WHEN wp_posts.post_content LIKE '%post format%' THEN 5 ELSE 6 END), wp_posts.post_date DESC, wp_posts.ID DESC 
LIMIT 0, 11
```

This is a straightforward example and it's easy for the error to go unnoticed. Forward and backward cursoring works fine, we're just missing results. We can see where this goes off the rails if we change the first/last amounts to 15. I forked the [pagination example](https://codesandbox.io/s/wpgraphql-apollo-forward-backward-pagination-example-spq53) to play this out and make a gif that illustrates everything I walk through below:

![wpgraphql-1818-example](https://user-images.githubusercontent.com/465154/220132185-36ac21a6-c963-42dc-9571-06953f379961.gif)

The first 15 are what we expect. But now the 15th post's date is newer than the 14 before it, so when we get the next set of results (should be after `YXJyYXljb25uZWN0aW9uOjExNzg=`), we end up seeing the first 14 repeated. The 15th post is actually not what we expect, we skipped over two newer posts in the data set from 2018 (generated SQL query below).

```sql
SELECT   wp_posts.ID
FROM wp_posts 
WHERE 1=1  AND (((wp_posts.post_title LIKE '%post%') OR (wp_posts.post_excerpt LIKE '%post%') OR (wp_posts.post_content LIKE '%post%')) AND ((wp_posts.post_title LIKE '%format%') OR (wp_posts.post_excerpt LIKE '%format%') OR (wp_posts.post_content LIKE '%format%')))  AND wp_posts.post_type = 'post' AND ((wp_posts.post_status = 'publish')) AND  CAST( wp_posts.post_date as DATETIME ) <= CAST( '2013-01-11 20:22:19' as DATETIME ) AND ( CAST( wp_posts.post_date as DATETIME ) < CAST( '2013-01-11 20:22:19' as DATETIME ) OR (  wp_posts.ID < 1178 ) ) 
ORDER BY (CASE WHEN wp_posts.post_title LIKE '%post format%' THEN 1 WHEN wp_posts.post_title LIKE '%post%' AND wp_posts.post_title LIKE '%format%' THEN 2 WHEN wp_posts.post_title LIKE '%post%' OR wp_posts.post_title LIKE '%format%' THEN 3 WHEN wp_posts.post_excerpt LIKE '%post format%' THEN 4 WHEN wp_posts.post_content LIKE '%post format%' THEN 5 ELSE 6 END), wp_posts.post_date DESC, wp_posts.ID DESC 
LIMIT 0, 16
```

If we try to go backwards, it gets worse. Because we'll be looking for newer posts than the first result, and the first result is a very old post, the date clause will remove the first 14 results and we're just looking at the last 5, but in an order I can't make much sense of -- if I expect those posts in order a-b-c-d-e, I see them instead in order b-c-d-e-a (I suspect that WPGraphQL is reversing the order of the results in SQL expecting to reverse it back later, but the "relevance" ordering is makes that reverse-back incorrect). Further, `pageInfo` indicates that there's a next page, which I also can't make heads or tails of, since the query asked for 16 posts and got 5. 🤷 Here's the last SQL query.

```sql
SELECT   wp_posts.ID
FROM wp_posts 
WHERE 1=1  AND (((wp_posts.post_title LIKE '%post%') OR (wp_posts.post_excerpt LIKE '%post%') OR (wp_posts.post_content LIKE '%post%')) AND ((wp_posts.post_title LIKE '%format%') OR (wp_posts.post_excerpt LIKE '%format%') OR (wp_posts.post_content LIKE '%format%')))  AND wp_posts.post_type = 'post' AND ((wp_posts.post_status = 'publish')) AND  CAST( wp_posts.post_date as DATETIME ) >= CAST( '2010-10-05 00:27:25' as DATETIME ) AND ( CAST( wp_posts.post_date as DATETIME ) > CAST( '2010-10-05 00:27:25' as DATETIME ) OR (  wp_posts.ID > 358 ) ) 
ORDER BY (CASE WHEN wp_posts.post_title LIKE '%post format%' THEN 1 WHEN wp_posts.post_title LIKE '%post%' AND wp_posts.post_title LIKE '%format%' THEN 2 WHEN wp_posts.post_title LIKE '%post%' OR wp_posts.post_title LIKE '%format%' THEN 3 WHEN wp_posts.post_excerpt LIKE '%post format%' THEN 4 WHEN wp_posts.post_content LIKE '%post format%' THEN 5 ELSE 6 END), wp_posts.post_date ASC, wp_posts.ID ASC 
LIMIT 0, 16
```

Let me know if you need any other info on this! Happy to provide whatever would be helpful.

**Comment by jasonbahl** [comment]
@mboynes thanks so much for the details on this! This should help us find a fix 🙏🏻

**Comment by benknight** [comment]
@mboynes thanks for adding context! Looks like we reached the same conclusions :)

**Comment by kamilgrzegorczyk** [comment]
Hello guys, thx for all the hard work on the plugin.

Is there any update on that issue? We seem to have the same issue when running the search through the ElasticPress where the cursor is also skipping a lot of posts and it is not consistent. 

The issue is gone when ElasticPress search is disabled and original WP search is used.

**Comment by alaa-alshamy** [comment]
I can confirm that the cursor pagination is failing in searching case, and that's because the relevance order is taking higher priority above WPGraphQL cursor order which is by date and ID, so the where filitering may miss some posts if the end cursor is older than the rest posts.

So the relevance order should be taking into your consideration like other orders here:
https://github.com/wp-graphql/wp-graphql/blob/develop/src/Data/Cursor/PostObjectCursor.php#L194

So is this something will be fixed soon ?

**Comment by justlevine** [comment]
[WP_Query::parse_search_order()](https://github.com/WordPress/wordpress-develop/blob/6.4/src/wp-includes/class-wp-query.php#L1590) is the core method responsible for handling this. 

I'm not entirely sure how best to translate it to our `PostObjectCursor` logic.

**Comment by catative** [comment]
I think the bug is related to the fact that you need to add sorting to the queries (for example, add `orderby: {field: NAME, order: ASC})`, then cursors work well. Perhaps this will be applicable not only to my case (using where), but also in queries with search. 

I had this problem: there was the following get request
```
query getProductsByCategorySlug($slug: [String], $first: Int, $after: String) {
  products(
    first: $first
    after: $after
    where: {categoryIn: $slug, typeNotIn: EXTERNAL, visibility: VISIBLE, status: "publish"}
  ) {
    pageInfo {
      hasNextPage
      endCursor
    }
    nodes {
      ... on SimpleProduct {
        id
        name
        slug
        uri
        image {
          altText
          mediaItemUrl
        }
        galleryImages {
          nodes {
            altText
            mediaItemUrl
          }
        }
        productCategories {
          nodes {
            name
            slug
          }
        }
      }
    }
  }
  productCategories(where: {slug: $slug}) {
    nodes {
      name
      slug
    }
  }
}
```

when I received the **endCursor** value after first request and set it into the **after** variable in the second request, the list of products was always returned the same. However, once I added sorting, everything started working well:
```
query getProductsByCategorySlug($slug: [String], $first: Int, $after: String) {
  products(
    first: $first
    after: $after
    where: {categoryIn: $slug, typeNotIn: EXTERNAL, visibility: VISIBLE, status: "publish", **orderby: {field: NAME, order: ASC}**}
  ) {
    pageInfo {
      hasNextPage
      endCursor
    }
    nodes {
      ... on SimpleProduct {
        id
        name
        slug
        uri
        image {
          altText
          mediaItemUrl
        }
        galleryImages {
          nodes {
            altText
            mediaItemUrl
          }
        }
        productCategories {
          nodes {
            name
            slug
          }
        }
      }
    }
  }
  productCategories(where: {slug: $slug}) {
    nodes {
      name
      slug
    }
  }
}
```
I was inspired by [this project](https://github.com/scottyzen/woonuxt) and in particular [this file](https://github.com/scottyzen/woonuxt/blob/master/woonuxt_base/app/queries/getProducts.gql)

**Comment by furai** [comment]
Any news on this? Or is only solution to apply sorting to the query to stabilise the cursor? @jasonbahl any ideas?

**Comment by jasonbahl** [comment]
Thanks everyone for the patience here, and especially @mboynes for the reproduction with theme unit test data. We've confirmed the root cause against the current codebase: the guard that was meant to force date ordering for searches checks `$query_args['search']`, but the arg is mapped to `s` before that check runs, so it has been dead code since it was added, and meanwhile the cursor layer knew nothing about the relevance ordering WP_Query applies to searches.

Rather than forcing searched results into date order (the trade-off that stalled #1819), the fix in #3935 makes the cursor relevance-aware, so search results stay ordered by relevance and now paginate correctly in both directions. Regression tests use a fixture where relevance and date order genuinely differ, which is the case the previous tests missed.

**Comment by benknight** [comment]
@jasonbahl I don't mean to be a poop but my original post in 2021 pinpointed the exact reason you just described! Anyway thanks for the fix.

**Comment by jasonbahl** [comment]
It was deeper than the "s" vs "search" mapping as cursor based pagination wasn't working with the relevance based sorting and previous attempts to resolve, from myself and others, caused search results to be ordered by date which wasn't ideal. The PR now includes fixes that map our cursor pagination to support relevance based sorting.
